### DISCUSSION TITLE
ipxe failed to fetch the tftp due to two dhcp interfaces

### DISCUSSION THREAD

**Discussion by simoncheel** [ipxe/ipxe#827] [General]
Dear developer
I'm struggling on this issue for quite sometime

My computer has two interfaces , and both connect to the network that have dhcp server. during the pxe boot, the NIC can get the ip address from bootp server, but when try to load the boot configuration file failed from tftp server (connection timed out).

I searched around, and do a lot of test, I found the regular pxe boot loader does not have this issue

and finally find that the ipxe will try to active the interfaces from the begining, and when the first nic actived, it will not proceed the further checking, that will introduce a problem when I have two interfaces that both have dhcp will asign the ip address

can you pls guide me, how to resolve the issue in such situtation

expect your repsonse, thanks in advance.

**Comment by stappersg** [comment]
> can you pls guide me, how to resolve the issue in such situtation

:smiley: 

> My computer has two interfaces , and both connect to the network that have dhcp server. during the pxe boot, the NIC can get the ip address from bootp server, but when try to load the boot configuration file failed from tftp server (connection timed out).

Start with using just one interface.


> expect your repsonse, thanks in advance.

Unaccepted, thank the iPXE project when you have a working iPXE setup.

**Comment by simoncheel** [comment]
that's not a good idea... I need to have two dhcp server

**Reply by stappersg** [reply]
> I need to have those two DHCP servers

Acknowledge on **your** needs, do know that it makes you the biggest stakeholder.


> that is no a good idea

Rejecting advice without considering it, is neither a good idea.
There was no request to stop one DHCP server forever.
The idea was and is to boot the iPXE client seeing/hearing only **one** DHCP server.

Please try it, please report back.

**Reply by simoncheel** [reply]
how to do this?
The idea was and is to boot the iPXE client seeing/hearing only one DHCP server.
now the client try to acquire the ip from the dhcp server which not on the deployment network.

**Reply by stappersg** [reply]
> Now the client tries to acquire an IP-address from the DHCP server which is not on the deployment network.

Make the iPXE client seeing/hearing the other DHCP server.

**Comment by NiKiZe** [comment]
iPXE tries each NIC, if it has no information about boot it will try the next NIC, make sure that only the one you want to boot from sends out next-server and filename.

This is not a bug or something that can be fixed in iPXE, however you might want to consider an embedded script with logic.

**Reply by simoncheel** [reply]
I enabled the debug, here is the procedure
it turns out the ipxe try to boot from net0 to search for ipxe.config file, but actually, it should be on the net4

How can I skip the first net0 and try to enable net4 for tftp file download?

![image](https://user-images.githubusercontent.com/119644964/205252054-5c8a1748-4623-4816-b5ca-1313e77dd237.png)
![image](https://user-images.githubusercontent.com/119644964/205252149-b8040813-52ff-4dde-b532-de45dcdc5f62.png)
![image](https://user-images.githubusercontent.com/119644964/205252236-6fcbd6f4-7672-4d19-939e-fe7b275a141a.png)

**Reply by simoncheel** [reply]
![image](https://user-images.githubusercontent.com/119644964/205252868-ac6add32-d0df-4cbf-8ec9-6131a421752e.png)

**Reply by simoncheel** [reply]
https://ipxe.org/cmd/ifconf
Automatically configure a network interface. iPXE will open the first specified network interface and attempt to automatically configure it. If automatic configuration succeeds, the command will terminate and the network interface will be left open. If automatic configuration fails, the network interface will be closed and iPXE will proceed to the next network interface in the list. 

here it says, if the first nic activated, it will not search for the others, I think I have this problem. net0 is not what I want, I need the net4 for ipxe boot

**Reply by simoncheel** [reply]
![image](https://user-images.githubusercontent.com/119644964/205255015-7267a7f9-718e-48e8-9746-c91e91c4f1e8.png)
you can see here, net0 and net4 both registered. and obtain the ip address from net4

**Reply by stappersg** [reply]
Repeating the hint from above
> however you might want to consider an embedded script

**Reply by NiKiZe** [reply]
Relevant command that runs internally on normal boot is 
https://ipxe.org/cmd/autoboot

And I repeat, don't have tftpserver or filename defined and boot should fail and move on to the next one.

But if you are chaining in to iPXE from pxe boot, then consider using snponly.efi instead so that you only get that one NIC that it was booted from, unless you need all of them.

Now to verify what happens, can you boot into iPXE, press ctrl-b to get into shell before any automatic configuration is done and verify with the config command that you have no boot information (filename or next-server)
Then run ifconf let it gets its IP, now again verify with config command what options you got.

After thinking about it, it might maybe be possible that iPXE gets cached information for these fields which makes the intended behaviour fail.

**Reply by simoncheel** [reply]
I also tried autoboot, and then compile it again, but still not able to boot ...

**Reply by NiKiZe** [reply]
Do you have an embedded script? If not then autobot is what runs by default.

You have still not verified however you get boot options on the network that you don't want. Because if you do then there is nothing to fix in iPXE, instead it is your setup.as also mentioned snponly.efi should only boot thru the NIC that it was chained from.

Unfortunately you seem to skip parts of the suggestion and requested information, which makes it hard, if not impossible to help you.

Please go back and read over the posts again. Verify your data, all of it, even if you are sure it is correct.
If you find the language difficult then you need to take extra care to not missunderstand.

**Reply by simoncheel** [reply]
tried snponly.efi, still not able to boot... 
![image](https://user-images.githubusercontent.com/119644964/206342729-6d082f25-dce3-48b4-afb0-d59d910b3634.png)

**Reply by stappersg** [reply]
https://ipxe.org/4c126092

**Comment by habetsm-xilinx** [comment]
You can usually configure the BIOS to boot from the NIC you want. You may need 2 aspects:
1. Configure the boot order to first PXE boot the slot you want. That may solve your issue. If not you may need step 2.
2. With some BIOSs you can disable PXE booting on a given slot. Disable that for the NIC you don't want to PXE boot. Do not disable the slot completely, as that will hide the NIC from the OS as well.

If none of this works for you, on some NICs you can disable PXE booting in the NIC setup itself. This is NIC specific.

**Reply by NiKiZe** [reply]
This will not help since ipxe.efi is booted, ipxe boots from "wanted" nic but then ipxe has native drivers for all the nics and does not know/care about bios order.

The fix here is to use snponly.efi that will only try to boot the nic that it was chained via.

**Reply by simoncheel** [reply]
I did that, but with no luck

**Reply by dfh0522** [reply]
> I did that, but with no luck

Did you try to configure dnsmasq with two interfaces on same host? this is my dnsmasq.conf

interface=eth1
enable-tftp
tftp-root=/tftproot
dhcp-boot=undionly.kpxe,b85m,192.168.1.11
#dhcp-host=10:c3:7b:96:92:9b,ignore
dhcp-host=eth1,48:df:37:22:9e:45,192.168.1.11,b85m
dhcp-host=eth1,48:df:37:1f:e5:08,192.168.1.12,asusb85

interface=eth2
dhcp-boot=undionly.kpxe,b85m,192.168.2.11
dhcp-host=eth2,48:df:37:22:9e:46,192.168.2.11,b85m
dhcp-host=eth2,48:df:37:1f:66:4e,192.168.2.9,bm1ad

It's nothing to do with ipxe.
